data/src/main/java/com/microsoft/azure/kusto/data/auth/TokenProviderBase.java

```diff
@@ -0,0 +1,104 @@
+    protected final String clusterUrl;
+    protected final Set<String> scopes;
+    protected final String aadAuthorityUrl;
+
+    TokenProviderBase(@NotNull String clusterUrl, String authorityId) throws URISyntaxException {
+        URI clusterUri = new URI(clusterUrl);
+        this.clusterUrl = String.format("%s://%s", clusterUri.getScheme(), clusterUri.getHost());
+        String scope = String.format("%s/%s", clusterUrl, ".default");
+        scopes = new HashSet<>();
+        scopes.add(scope);
+
+        aadAuthorityUrl = determineAadAuthorityUrl(authorityId);
+    }
+
+    private String determineAadAuthorityUrl(String authorityId) {
+        if (authorityId == null) {
+            authorityId = ORGANIZATION_URI_SUFFIX;
+        }
+
+        String authorityUrl;
+        String aadAuthorityFromEnv = System.getenv("AadAuthorityUri");
+        if (aadAuthorityFromEnv == null) {
+            authorityUrl = String.format("https://login.microsoftonline.com/%s/", authorityId);
+        } else {
+            authorityUrl = String.format("%s%s%s/", aadAuthorityFromEnv, aadAuthorityFromEnv.endsWith("/") ? "" : "/", authorityId);
+        }
+        return authorityUrl;
+    }
+
+    public String acquireAccessToken() throws DataServiceException, DataClientException {
+        IAuthenticationResult accessTokenResult = acquireAccessTokenSilently();
+        if (accessTokenResult == null) {
+            accessTokenResult = acquireNewAccessToken();
+        }
+        return accessTokenResult.accessToken();
+    }
+
+    protected IAuthenticationResult acquireAccessTokenSilently() throws DataServiceException, DataClientException {
+        try {
+            // TODO: MSAL docs say this would throw MsalUiRequiredException, but it doesn't in Java
+            return acquireAccessTokenSilentlyMsal();
+        } catch (MalformedURLException e) {
+            throw new DataClientException(clusterUrl, ERROR_INVALID_AUTHORITY_URL, e);
+        } catch (TimeoutException | ExecutionException e) {
+            return null; // Legitimate outcome, in which case a new token will be acquired
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            return null; // Legitimate outcome, in which case a new token will be acquired
+        }
+    }
+
+    protected IAuthenticationResult acquireAccessTokenSilentlyMsal() throws MalformedURLException, InterruptedException, ExecutionException, TimeoutException, DataServiceException {
+        throw new DataServiceException("Cannot obtain a refresh token for Authentication type '" + this.getClass().getSimpleName() + "'");
+    }
+
+    protected IAuthenticationResult acquireNewAccessToken() throws DataServiceException, DataClientException {
+        throw new DataServiceException("Cannot obtain a new token for Authentication type '" + this.getClass().getSimpleName() + "'");
+    }
+
+    SilentParameters getSilentParameters(Set<IAccount> accountSet) {

```

Comment:
Can you unify this and the below functions into one?

Reply:
They used to be unified, but I split them in order to allow just getAccount() in UserPromptTokenProvider to be overriden, while reducing code duplication.

Reply:
See? 
You need base and MSAL Base, and then you will not need to jump over hoops to make the design work

Reply:
i.e. you should implement functionality in rather then override functionality out

Reply:
I don't disagree with adding an additional layer of abstraction, but just to point out that this isn't an example of overriding functionality out, since getSilentParameters isn't overrided out. It's just not *used* in the TokenProvider and Token cases.

But I agree with the pros for another layer of abstraction, and though I hadn't done so originally to avoid making the design overly complex, I think you're right that it's worth it.

Reply:
I chose to avoid a 3 level hierarchy as well, but at the cost of repeating myself on the different MSAL implementations.
i.e. my base impl is very very simple and contains abstract functions for get token and get token silent.
If you want to go by the book though, use 3 levels (base, msal base, actual impl of base or msal base)